ocs_ci/ocs/version.py

```diff
@@ -0,0 +1,179 @@
+    # instance. See also:
+    # https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusterversion.md
+    ocp = OCP(kind="clusterversion")
+    cluster_version = ocp.get("version")
+
+    # TODO: When OLM (operator-lifecycle-manager) maintains OCS, it will be
+    # possible to add a check via CSV (cluster service version) to get OCS
+    # version in a similar way as it's done for OCP itself above.
+    # Reference: Jose A. Rivera on ocs-qe list.
+
+    # Get all openshift storage related namespaces. Eg. at the time of writing
+    # this code (July 2019), there were these storage namespaces:
+    #  * openshift-cluster-storage-operator
+    #  * openshift-storage
+    # TODO: how to do this in upstream where namespace is rook-ceph?
+    # TODO: should we use defaults.ROOK_CLUSTER_NAMESPACE somehow here?
+    storage_namespaces = []
+    all_namespaces = OCP(kind="namespace").get()['items']
+    ns_re = re.compile("^openshift.*storage")
+    for ns in all_namespaces:
+        if (ns_re.match(ns['metadata']['name'])):
+            storage_namespaces.append(ns['metadata']['name'])
+
+    logger.info("found storage namespaces %s", storage_namespaces)
+
+    # Now get the OCS version by asking for version of all container images of
+    # all pods in openshift-storage namespace.
+    image_dict = {}
+    for ns in storage_namespaces:
+        ocs = OCP(kind="pod", namespace=ns)
+        ocs_pods = ocs.get()
+        ns_dict = {}
+        for pod in ocs_pods['items']:
+            for container in pod['spec']['containers']:
+                logger.debug(
+                    'container %s (in pod %s) uses image %s',
+                    container['name'],
+                    pod['metadata']['name'],
+                    container['image'])
+            cs_items = pod['status'].get('containerStatuses')
+            if cs_items is None:
+                pod_name = pod['metadata']['name']
+                logger.warning(f'pod {pod_name} has no containerStatuses')
+                continue
+            for cs in cs_items:
+                ns_dict.setdefault(cs['image'], set()).add(cs['imageID'])
+        image_dict.setdefault(ns, ns_dict)
+
+    logger.debug("ocs version collected")
+
+    return cluster_version, image_dict
+
+
+def report_ocs_version(cluster_version, image_dict, file_obj):
+    """
+    Report OCS version via:
+
+     * python logging
+     * printing human readable version into file_obj (stdout by default)
+    """
+    # log the version
+    logger.info("ClusterVersion .spec.channel: %s", cluster_version["spec"]["channel"])
+    logger.info("ClusterVersion .status.desired.version: %s", cluster_version["status"]["desired"]["version"])
+    logger.info("ClusterVersion .status.desired.image: %s", cluster_version["status"]["desired"]["image"])
+
+    # write human readable version of the above
+    print(f'cluster channel: {cluster_version["spec"]["channel"]}', file=file_obj)
```

Comment:
Is there a reason we are printing as well as logging some things? Would it be better for us to log consistently?

Reply:
The printing lines are used to create human readable plain text report, which is placed into a file (if this is executed during deployment), or could be printed to stdout if you run it via a command line tool. This plaintext report could be copy pasted into bugreport, and through it, I can get some basic idea what OCS components and in which versions have been deployed. Example:

```
cluster channel: stable-4.2
cluster version: 4.2.0-0.nightly-2019-12-02-165545
cluster image: registry.svc.ci.openshift.org/ocp/release@sha256:159ba31ba77bc5ae78d09caca06eada94483b27f4f8e8afafdfa036b42bd3d80

storage namespace openshift-cluster-storage-operator
image quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1f2819a25c4f1945d3a2a63e64b89a4f1f095cbb34e39829785e966151340fdc
 * quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1f2819a25c4f1945d3a2a63e64b89a4f1f095cbb34e39829785e966151340fdc

storage namespace openshift-storage
image quay.io/rhceph-dev/cephcsi:4.2-225.72ac53b6.release_4.2
 * quay.io/rhceph-dev/cephcsi@sha256:69950f3c8bdf6889e85deafdbc8d07e63910337435d2f5d6a1ba674d08f85c51
image quay.io/openshift/origin-csi-node-driver-registrar:4.2
 * quay.io/openshift/origin-csi-node-driver-registrar@sha256:6671a4a02a9bf5abfa58087b6d2fea430278d7bc5017aab6b88a84203c0dad06
image quay.io/openshift/origin-csi-external-attacher:4.2
 * quay.io/openshift/origin-csi-external-attacher@sha256:dafeaa49292fce721c1ed763efd1f45b02a9c4f78d7087a854067c5e2718d93c
image quay.io/openshift/origin-csi-external-provisioner:4.2
 * quay.io/openshift/origin-csi-external-provisioner@sha256:dbe8b5e1bebfed1e8f68be1d968ff0356270a0311c21f3920e0f3bec3b4d97ea
image quay.io/openshift/origin-csi-external-snapshotter:4.2
 * quay.io/openshift/origin-csi-external-snapshotter@sha256:910f54893ac1ae42b80735b0def153ee62aa7a73d5090d2955afc007b663ec79
image quay.io/rhceph-dev/mcg-core:5.2.11-20.59b35a4f0.5.2
 * quay.io/rhceph-dev/mcg-core@sha256:3f9da7525a751837750cfe8fd49d8707090c7a1c4e257c23b882184e5f97f76e
image registry.access.redhat.com/rhscl/mongodb-36-rhel7:latest
 * registry.access.redhat.com/rhscl/mongodb-36-rhel7@sha256:c845fa631fafab54addb20e33af2a8bc2a693af07ffd9a9fa1bc2715d39b40ca
image quay.io/rhceph-dev/mcg-operator:2.0.9-35.db6a79b.2.0
 * quay.io/rhceph-dev/mcg-operator@sha256:e065c52e0132450477223f4262a495db84de0963ed639457c961e52c36618f35
image quay.io/rhceph-dev/ocs-operator:4.2-259.e29a26e.release_4.2
 * quay.io/rhceph-dev/ocs-operator@sha256:339ad733e127e7eb182e46cfd74200615c42bd4f1d44331e2f13479b8ccfad8d
image quay.io/rhceph-dev/rook-ceph:4.2-257.9a1d8592.ocs_4.2
 * quay.io/rhceph-dev/rook-ceph@sha256:1ae7998e77ccea9a07a39da0d196326d9e485d05bd24df48c088a740dd219410
image quay.io/rhceph-dev/rhceph:4-119.20191128.ci.0
 * quay.io/rhceph-dev/rhceph@sha256:2bbb7704bae17b8a20c6ff1c240d2b1e83cd185fff2aa80ccc85968365cbb4e0
```

On the other hand, the logging is there to make sure that minimal version info is present in every log file, in case this information is needed. The format of this version logging could be definitely tweaked.

Reply:
I see, so the human readable output still ends up in a file? What file is that? My only real concern was that we would be logging this to our log files but the human readable output would have to be dug out of the console output.

Reply:
The file is `CLUSTER_DIR/ocs_version.TIMESTAMP` when this is executed via run-ci.